GNINA molecular docking with CNNs

ncfrey opened the pull request:
## Description

This PR address #2372, adding support for the new stable release of [GNINA](https://github.com/gnina/gnina) through DeepChem's docking API.  GNINA provides CNN docking scores that outperform AutoDock Vina for pose generation and virtual screening. The ability to run GNINA on GPUs and train new CNN scoring models also brings the docking module more in line with DeepChem's core deep learning focus. 

Thanks to @dkoes, there is a binary distribution with the [ v1.0 release](https://github.com/gnina/gnina/releases/tag/v1.0) that makes installation of GNINA very quick on Colab and most local machines with CUDA>=10.1.


## Type of change

- [X] New feature (non-breaking change which adds functionality)

CCing @dkoes @rbharath @peastman who may be interested

peastman commented:
If this feature is only supported on Linux, we should say that in the documentation, and also make it skip the tests on other platforms.  You have it skipping one of the tests on Windows, but not the other one, and it doesn't skip either of them on MacOS.

Is it possible to support this on other platforms?  They only have a precompiled binary for Linux, but they recommend people compile it themselves rather than use that version, and I couldn't find any mention of whether it works on Mac or Windows.  Is there a way we can let people use a binary they compiled themselves?

If vina_utils.py is going to include functions for gnina, perhaps we should rename it to docking_utils.py?

dkoes commented:
I would need someone to contribute a Windows build.  Building on Mac requires CUDA, which isn't impossible but you'll never actually have GPU acceleration.

rbharath commented on deepchem/dock/pose_generation.py:
```diff
@@ -16,7 +16,7 @@
 from deepchem.utils.geometry_utils import compute_centroid, compute_protein_range
 from deepchem.utils.rdkit_utils import load_molecule, write_molecule
-from deepchem.utils.vina_utils import load_docked_ligands, write_vina_conf
+from deepchem.utils.vina_utils import load_docked_ligands, write_vina_conf, write_gnina_conf, read_gnina_log
```
+1 for @peastman's suggestion to rename this `deepchem.utils.docking_utils`.

On a more general note, we should figure out our policy with API stability on utils files. To be safe, we could make the shift backwards compatible by having wrappers in `vina_utils` that emit deprecation warnings.

rbharath commented on deepchem/dock/pose_generation.py:
```diff
@@ -79,6 +79,166 @@ def generate_poses(self,
+class GninaPoseGenerator(PoseGenerator):
+  """Use GNINA to generate binding poses.
+
+  This class uses GNINA (a deep learning framework for molecular
```
This is excellent documentation!

rbharath commented on deepchem/dock/pose_generation.py:
```diff
@@ -79,6 +79,166 @@ def generate_poses(self,
+
+  Notes
+  -----
+  * GNINA requires CUDA >= 10.1 for fast CNN scoring.
```
Can we add a note here that Gnina currently only works on linux?

rbharath commented on deepchem/dock/pose_generation.py:
```diff
@@ -79,6 +79,166 @@ def generate_poses(self,
+  "Protein–Ligand Scoring with Convolutional Neural Networks."
+  Journal of chemical information and modeling (2017).
+
+  Notes
```
Can you change this to "Note" instead? That makes the readthedocs render a nice note blurb

rbharath commented on deepchem/dock/pose_generation.py:
```diff
@@ -79,6 +79,166 @@ def generate_poses(self,
+      self.gnina_cmd = os.path.join(self.gnina_dir, filename)
+    else:
+      raise ValueError(
+          "Unknown operating system. Try using a cloud platform to run this code instead."
```
Perhaps "Gnina currently only runs on Linux. Try using a cloud platform to run this code instead."

rbharath commented on deepchem/utils/test/data/gnina_log.txt:
```diff
@@ -0,0 +1,25 @@
+              _             
```
As a quick question for @dkoes, can we assume that this log file format is stable? The PR assumes that the log file will follow this set structure to read/write log files gnina generates.

dkoes replied:
I can't guarantee it in perpetuity, but there are no changes planned that would invalidate the processing.  If you want a more robust solution, extract the scores from the output sdf.

rbharath reviewed (commented):
This is excellent! I have a few minor comments below

ncfrey commented:
I addressed the comments above, with the biggest changes being skipping tests that aren't on Linux and deprecating `vina_utils.` To retain compatibility with Python3.6, I used a `@deprecated` decorator function. Happy to use a different implementation if there is a preferred way to deprecate the module.

peastman commented:
I think we want to just rename vina_utils.py to docking_utils.py.  Keeping around duplicate copies of the functions is a great opportunity for bugs!

rbharath commented on deepchem/utils/vina_utils.py:
```diff
@@ -56,84 +78,7 @@ def write_vina_conf(protein_filename: str,
-
-
+@deprecated
 def load_docked_ligands(
```
Perhaps we could import the functions from `docking_utils.py` and then have only deprecation wrappers in `vina_utils.py` so we don't have duplicate code?

ncfrey replied:
Got it, will do!

rbharath reviewed (commented):
Looking good overall! +1 to @peastman's comment and a suggestion below on how to implement it

ncfrey commented:
I removed the duplicate code as suggested by @peastman, following @rbharath's suggestion to wrap the functions in `vina_utils`.

rbharath reviewed (approved):
This PR is looking good! I think this is ready to merge in on my end :)

 Will leave open in case for a bit in case other folks still have comments

ncfrey commented:
The same pymatgen-related errors are showing up here, as in #2386. I think this is ok to merge, but I'll open a new issue to note that these errors need to be fixed.

ncfrey commented:
Other than docs, this PR is now passing all tests and should be ready to merge.

rbharath commented:
Great! Going to merge this in
